codegen/codegen.go

```diff
@@ -30,31 +34,29 @@ func (c *CodeGenerator) GenBinding(binding *ast.Binding) {
 	// TODO: generate function template
 }
 
-func (c *CodeGenerator) CallBindingWith(builder llvm.Builder, binding *ast.Binding, vs []llvm.Value) llvm.Value {
+func (c *CodeGenerator) CallBindingWith(builder Builder, binding *ast.Binding, vs []value.Value) value.Value {
 	if len(vs) == len(binding.ParamList) {
-		typeList := make([]llvm.Type, len(vs))
-		scope := map[string]llvm.Type{}
-		valueScope := map[string]llvm.Value{}
+		params := make([]*ir.Param, len(vs))
+		scope := map[string]types.Type{}
+		valueScope := map[string]value.Value{}
 		for i, v := range vs {
 			paramName := binding.ParamList[i]
 			scope[paramName] = v.Type()
 			valueScope[paramName] = v
-			typeList[i] = v.Type()
+			params[i] = ir.NewParam(paramName, v.Type())
 		}
 		retT := c.GetExprType(scope, binding.Expr)
-		ft := llvm.FunctionType(retT, typeList, false)
-		newFn := llvm.AddFunction(c.mod, binding.Name, ft)
-		newFunctionBuilder := llvm.NewBuilder()
-		block := llvm.AddBasicBlock(newFn, "")
-		newFunctionBuilder.SetInsertPointAtEnd(block)
+		newFn := c.mod.NewFunc(binding.Name, retT, params...)
+		block := newFn.NewBlock("")
+		newFunctionBuilder := block
```

Comment:
`newFnBuilder := newFn.NewBlock("")`

Reply:
> `newFnBuilder := newFn.NewBlock("")`

Done.

Reply:
I made the update, but after thinking about it further, there is a benefit with having the basic block created explicitly (as was done in the original code). The reason is that you may want to update how the builder works, right now, it always uses `SetInsertPointAtEnd` by default.

In particular, I would imagine that the [type Builder interface](https://github.com/mewpull/elz/blob/6d3a3e1c1004e2a57ec9052d39cdf7cb08ac1071/codegen/builder.go#L11) in builder.go would be replaced with a concrete builder implementation that allows for more flexible control over insertion points for instructions.

This is just a suggestion however, and you can update the code at that time, should the need arise.

Cheers,
Robin

Reply:
Thanks for the explanation, I'm also thinking why not use concrete type XD.